uniform gaps

BertVanRaemdonck opened the pull request:
With the new gap creating algorithm, every brick shrinks a fixed amount in every direction, no matter what scale of dimensions it has. this results in uniform gaps, in contrast to when you would just rescale the object uniformly: then you get bigger gaps in the direction the part is longer.

le717 commented:
Linking back to #95 for reference.

le717 commented:
Thanks for the PR! I'll take a quick look at this now, but I need to ask if you can name a few parts I can test this on to ensure this is a good change to perform.

le717 commented on import_ldraw.py:
```diff
@@ -820,7 +820,25 @@ def create_model(self, scale, context):
+                    # the dimension so that the mesh shrinks a fixed distance
+                    # (determined by the gap_width and the scale of the object)
+                    # in every direction, creating a uniform gap.
+                    if dim.x != 0: fac_x = 1-gap_width*abs(obj_scale.x)/dim.x
```
Although it may look a bit nicer this way, please put the statements on a new line as the Python style guide says.

le717 commented on import_ldraw.py:
```diff
@@ -820,7 +820,25 @@ def create_model(self, scale, context):
+                    # in every direction, creating a uniform gap.
+                    if dim.x != 0: fac_x = 1-gap_width*abs(obj_scale.x)/dim.x
+                    else: fac_x = 1
+                    if dim.y != 0: fac_y = 1-gap_width*abs(obj_scale.y)/dim.y
```
Needs spaces between the operators.

le717 commented on import_ldraw.py:
```diff
@@ -820,7 +820,25 @@ def create_model(self, scale, context):
                     # Add small gaps between each brick
-                    bpy.ops.transform.resize(value=(0.99, 0.99, 0.99))
+                    gap_width = 0.3
+                    obj_scale = cur_obj.scale * scale
```
The script (mostly) uses `camelCase` format, but with the shorter names you have here (e.g., `fac_x`), perhaps these should remain using `under_scores`. I'll have to think about that.

BertVanRaemdonck commented:
Thanks for looking at it. I'm not too familiar with these styles of codes, as I'm just a casual programmer.  I just copy-pasted this out of my own version. And just today I was thinking maybe it isn't necessary to get the object's scale, as that will always be equal to one unless something's changed elsewhere in the program. That way the code would be just a bit faster.

As for bricks to check on, you can use all bricks in principle. I'd suggest building a wall using several sizes of normal bricks to appreciate the seams, or take a large plate and tile it with 1x8 tiles. That's where I noticed the problem with the previous version: large gaps on the short ends, small gaps on the long ends.

le717 commented:
> Thanks for looking at it. I'm not too familiar with these styles of codes, as I'm just a casual programmer. I just copy-pasted this out of my own version.

No problem, that's why I read over code and point out errors. ;)

> And just today I was thinking maybe it isn't necessary to get the object's scale, as that will always be equal to one unless something's changed elsewhere in the program. That way the code would be just a bit faster.

There is an scale option in the import settings ([see this section](https://github.com/le717/LDR-Importer/blob/b998d94c40d0f86108be9faa09c1471825c0c98c/import_ldraw.py#L741-L750)), so that must be taken into consideration.

I'm running this code right now, and if all checks out I'll merge it. :)

BertVanRaemdonck commented:
As for that scaling, you see that it computes the resize values from the "obj_scale" variable. That on its turn depends on the actual scale of the object and the scaling factor you can set (the one you are referring to). As far as I know the first one of those two will always be 1, as that is the default and the object aren't scaled. The global scale factor, on the other hand, is very important as it largely determines the object's dimensions.

BertVanRaemdonck commented:
I downloaded the latest version of the importer yesterday, and noticed that the code I wrote to generate uniform gaps wasn't working anymore. Making sure the right objects were selected in the loop did the trick. But still, when the "Link identical parts" option was enabled, the gaps weren't right. The reason was that the orientations are defined differently now, so a simple scale transform doesn't do the trick anymore. It works fine by setting the scale value of the object itself. This doesn't change the behavior of bricks that aren't linked. Finally, I noticed my formula for calculating the factors was missing a factor of 2. This doesn't really change anything, but it helps to keep it physically connected. On my copy, everything seems to be working fine again :)

le717 commented:
If it isn't working in the latest release, that's because I never actually got around to merging this PR containing your gaps algorithm. :oops: I'll try to add it to my TODO list either this week or next.

le717 commented:
One thing I'd like to point out is that linked parts is technically broken. I allowed it to be built upon faulty logic and after the faulty foundation was fixed, linked bricks broke and I haven't looked at it yet. That might be something to be aware of with your updates.

BertVanRaemdonck commented:
Anyway, my code works for both linked and unlinked bricks, so there's no problem whatsoever.

Just out of curiosity: what's wrong with the linked parts? They work just like expected with me (more or less actually: when I go in edit mode and edit one part, it doesn't seem to affect the rest, but when I go back to object mode, the changes are copied to the other instances, so that's good). I also wonder why instances of the same geometry but different materials aren't linked. It's a bit strange that you edit a brick, but only the bricks with the same material follow... I modified the code a bit to link all parts with the same geometry, regardless of the material, and that's working just fine for me.

le717 commented:
It's broken in the sense that it was built on faulty logic and I let it happen. Upon fixing it I had to kludge the linked parts back together (b585f041d9688918c4bbda72480b85cbba67fcad). It needs to be fixed to work without that kludge, and that might mess up your code (IDK).

le717 commented on import_ldraw.py:
```diff
@@ -836,21 +836,41 @@ def create_model(self, context, scale):
+                if dim.z != 0:
+                    facZ = 1 - 2 * gapWidth * abs(objScale.z) / dim.z
+                else:
+                    facZ = 1
```
I would engineer this a tad differently.

* Create a `fac` dictionary with the default values:
```python
fac = {"x": 1, "y": 1, "z": 1}
```
* Update the values in the dictionary in the `!= 0` conditions, and drop the else blocks
* Revise the multiplication below to use the values in the dictionary (`*= fac["x"]`)

le717 commented on import_ldraw.py:
```diff
@@ -836,21 +836,41 @@ def create_model(self, context, scale):
+                bpy.context.object.scale[1] *= facY
+                bpy.context.object.scale[2] *= facZ
 
+                bpy.ops.object.transform_apply(location=False, rotation=False,\
```
The backslash is unnecessary. Also, because the first two parameters are named and [default to False](https://www.blender.org/api/blender_python_api_2_76_release/bpy.ops.object.html?highlight=transform_apply#bpy.ops.object.transform_apply), you should be able to simply reduce this line to

```python
bpy.ops.object.transform_apply(scale=True)
```

le717 commented:
I took at look and this and I am good for merging it. Once you get those changes made, let me know and I'll help you fix your commits. :)

BertVanRaemdonck commented:
I implemented the reasonable changes you requested, renamed the fac variable to scaleFac for clarity and changed a word here and there in the comments to make them a bit clearer. I'm glad you're liking it :)

Sorry I made such a mess with my commits. This is the first time I'm working with gitbash, setting up a remote etc., so I had to experiment a bit.

le717 commented:
Nice! I'll help you with Git after work. :)

le717 commented on import_ldraw.py:
```diff
@@ -836,21 +836,36 @@ def create_model(self, context, scale):
-                        # Go back to object mode
-                        bpy.ops.object.mode_set(mode='OBJECT')
 
+                gapWidth = 0.007    # Effectively controls the width of the gaps
```
I personally discourage the use of inline comments but I guess this is OK.

le717 commented:
OK, we want to flatten these commits into a single commit containing only the latest changes. Here's how:
* Open gitbash and run `git rebase -i HEAD~8` (8 because we want to flatten the last 8 commits)
* Set the state of each commit (`squash` or `fixup` for all except the newest, which would be `edit` so you can revise the commit message to be more accurate).
* When that is done and you can confirm the commit is correct, run `git push -f`. The `-f` does a force push, which is needed since you already pushed to the server.

BertVanRaemdonck commented:
Hmm... I'm running into some issues rebasing these commits. To begin with, when I enter `git rebase -i HEAD~9` (because now I've added an extra commit for that line comment), the file that opens contains no less than 59 commits, including ones that don't belong to me but to the original project... It goes all the way back to the creation of my branch. I guess that's because I've only made my branch a remote a couple of days ago, so all of the changes made to the original made in the meantime get in there... If I only go back to just before the `update` commit, in which I made my repo a remote of this one and merged it, there are no problems. What do you want me to do then?

le717 commented:
IDK how you updated it, but the way to update PRs with the latest changes it is pointed to is done by running `git merge upstream/<branch-name>`.

I am not quite sure what to do honestly. A quick fix is to grab the code, make a new branch containing the latest changes from `master`, merge the changes, then submit a new PR. I have no problem with that. :)
